Documentation on the parameters for the Config Object (#368)
columnflow/columnflow: docs/user_guide/building_blocks/config_objects.md

[the Analysis, Campaign and Config section](https://python-order.readthedocs.io/en/latest/quickstart.html#analysis-campaign-and-config)
of the Quickstart section of the order documentation.

After defining your Analysis object and your Campaign object(s), you can use the command
```python
cfg = analysis.add_config(campaign, name=your_config_name, id=your_config_id)
```
to create the
new Config object `cfg`, which will be
associated to both the Analysis object and the Campaign object needed for its creation. As the
Config object should contain the analysis-dependent information related to a certain campaign, it
should contain most of the information needed for running your analysis. Therefore, in this section,
the Config parameters required by Columnflow and some convenience parameters will be presented.

To start your analysis, do not forget to use the already existing analysis template in the
`analysis_templates/cms_minimal` Git directory and its
[config](https://github.com/columnflow/columnflow/blob/master/analysis_templates/cms_minimal/__cf_module_name__/config/analysis___cf_short_name_lc__.py).

> The Config saves informations under two general formats: the objects

Author: done

from the order package, which are necessary for your analysis to run in columnflow, and the
additional parameters, which are saved under the auxiliary key, accessible through the "x" key.
Unlike the general order package objects and while several names for parameters in the auxiliary
field do already have a meaning in columnflow and should respect the format used in columnflow,
the auxiliary field can contain any kind of

> parameter the user wants to save and reuse for parts of the analysis. These two general formats are

Author: trying to reformulate, is it good like that?

presented below.

> It is generally advised to create functions to separate the different parts of the filling of the Config object, especially as some parts will not change very often once the analysis is set up (e.g. luminosity), while others will be changed quite often, e.g. each time new variables are created. An example of such a separation can be found in the existing [hh2bbtautau analysis](https://github.com/uhh-cms/hh2bbtautau).

Author: done

## Parameters from the order package (required)

### Processes

> The physical processes to be investigated in the analysis. These should be saved as objects of the

Author: done

{external+order:py:class}`order.process.Process` class and added to the Config object using its

> {external+order:py:meth}`order.config.Config.add_process()` method. The processes added to the

Author: done, to check if it worked

> {external+order:py:meth}`order.config.Config.add_process()` method. The processes added to the

Author: checked

Config must correspond to the

> processes defined for the datasets and added to the Campaign object associated to the Config.

Author: done

It is possible to get all root processes from a specific campaign using the
{py:func}`~columnflow.config_util.get_root_processes_from_campaign()` function from columnflow.
Examples of information carried by a process could be the cross section of the process, registered

> under the {external+order:py:attr}`order.process.Process.xsecs` attribute and further used for normalization_weights in columnflow, and a color for the plotting scripts, which can be set using

Author: done

the {external+order:py:attr}`order.mixins.ColorMixin.color1` attribute of the process.
An example of a Process definition is given in the {ref}`Analysis, Campaign and Config <analysis_campaign_config>` section of the columnflow documentation.

> More informations about processes can be found in the

Author: done

{external+order:py:class}`order.process.Process` and the
{external+order:doc}`quickstart` sections of the order documentation.

### Datasets

The actual datasets to be processed in the analysis. These should be saved as objects of the
{external+order:py:class}`order.dataset.Dataset` class and added to the Config object using its
{external+order:py:meth}`order.config.Config.add_dataset()` method. The datasets added to the
Config object must correspond to the
datasets added to the Campaign object associated to the Config object. They are accessible through

> the {external+order:py:meth}`order.config.Campaign.get_dataset()` method of the Campaign class. The

Author: literalinclude added, to check if it worked

> the {external+order:py:meth}`order.config.Campaign.get_dataset()` method of the Campaign class. The

Author: checked

> Dataset objects should contain for example informations on the number of files and number of events

Author: done

> present in a Dataset as well as its key and if it is data or a Monte-Carlo dataset. It is also

Author: well this is the name of the argument, so I think we should use the word key, but I can add the description from the order documentation: "keys denote the identifiers or origins of a dataset" and add a note on get_dataset_lfns. Does it work for you like this?

> present in a Dataset as well as its key and if it is data or a Monte-Carlo dataset. It is also

Author: done

> present in a Dataset as well as its key and if it is data or a Monte-Carlo dataset. It is also

Author: done

> possible to change informations of a dataset in the config script. An example would be reducing the

Author: done

number of files to process for test purposes in a specific test config. This could be done with the
following lines of code:
e.g.
```python
n_files_max = 5
for info in dataset.info.values():
    info.n_files = min(info.n_files, n_files_max)
```

Once the processes and datasets have both been added to the config, one can check that the root
process of all datasets is part of any of the registered processes, using the columnflow function

> {py:func}`~columnflow.config_util.verify_config_processes()`.

Author: done

An example of a Dataset definition is given in the {ref}`Analysis, Campaign and Config <analysis_campaign_config>` section of the columnflow documentation.

### Variables

In order to create histograms out of the processed datasets, columnflow uses
{external+order:py:class}`order.variable.Variable`s. These
Variables need to be added to the config using the
function {external+order:py:meth}`order.config.Config.add_variable`. The standard syntax is as

> follows for `cfg` the Config object:

Author: done

```python

> cfg.add_variable(

Author: done for the change, not quite sure how to add it from the example config, as the lines go like this: 

```
cfg.add_variable(
    name="jets_pt",
    expression="Jet.pt",
    binning=(40, 0.0, 400.0),
    unit="GeV",
    x_title=r"$p_{T} of all jets$",
)
cfg.add_variable(
    name="jet1_pt",
    expression="Jet.pt[:,0]",
    null_value=EMPTY_FLOAT,
    binning=(40, 0.0, 400.0),
    unit="GeV",
    x_title=r"Jet 1 $p_{T}$",
)
```

Do you know a way to select only the variable I am interested in? I could give the lines but this might change for every change in the config, so I would rather not do that...

> cfg.add_variable(

Author: As far as I am aware, this works only on the first and last line of what you want to include, but this would be 
`cfg.add_variable(` and `)`, which is repeated for every variable and would then probably only take the first one, while I would prefer to have another one as example, e.g. with a null value.

> cfg.add_variable(

Author: added comments in template, now done

> cfg.add_variable(

Reviewer: Thank you for this update! I'm sorry I'm nagging so much, but:
- I wouldn't write the pseudo code you've included with the same argumentation as before - this is not testable and might be outdated at some point. Alternatively, you could e.g. add comments in the template that describe the different keyword arguments in more detail.
- I think there is a `)` missing in the include. Can you fix this? I.e. either try the `end-after` keyword or add an additional comment to use as an achor point

# Config Objects

## Generalities

The [order](https://github.com/riga/order) package defines several classes to implement the
metavariables of an Analysis. The order documentation and its {external+order:doc}`quickstart` section
provide an introduction to these different classes. In this section we will concentrate on the use
of the order classes to define your analysis.

> The three classes main classes needed to define your analysis are

Reviewer 2: ```suggestion
The three main classes needed to define your analysis are
```

Author: done

## Parameters from the order package (required)

### Processes

The physical processes to be investigated in the analysis. These should be saved as objects of the
{external+order:py:class}`order.process.Process` class and added to the Config object using its
{external+order:py:meth}`order.config.Config.add_process()` method. The processes added to the
Config must correspond to the
processes defined for the datasets and added to the Campaign object associated to the Config.
It is possible to get all root processes from a specific campaign using the
{py:func}`~columnflow.config_util.get_root_processes_from_campaign()` function from columnflow.
As an example of information carried by a process, a color for the plotting scripts can be set using
the {external+order:py:attr}`order.mixins.ColorMixin.color1` attribute of the process.
More informations about processes can be found in the
{external+order:py:class}`order.process.Process` and the

> {external+order:doc}`quickstart` sections of the order documentation.

Reviewer 2: It might be nice to mention that processes also carry the cross section information and that we can use that to determine normalization weights, linking to our normalization_weights Producer as an example

Reviewer 2: Also, it might be nice to link to some example in the cmsdb to showcase how the implementation of a process might look like (same for datasets)

Author: examples from the order docu are linked, I add a link to the cmsdb example provided in structure.md. Cross section information has been added

> An example with the transverse momentum of the first jet would be:

Author: done

With the ColumnCollection class, the example above could become:

```{literalinclude} ../../../analysis_templates/cms_minimal/__cf_module_name__/config/analysis___cf_short_name_lc__.py
:language: python
:start-after: "# columns to keep after certain steps"
:end-before: "# event weight columns as keys in an OrderedDict, mapped to shift instances they depend on"
```

<!--
```python
cfg.x.keep_columns = DotDict.wrap({
    "cf.ReduceEvents": {
        # general event info, mandatory for reading files with coffea
        ColumnCollection.MANDATORY_COFFEA
        # object info
        "Jet.pt", "Jet.eta", "Jet.phi", "Jet.mass", "Jet.btagDeepFlavB", "Jet.hadronFlavour",
        "Muon.pt", "Muon.eta", "Muon.phi", "Muon.mass", "Muon.pfRelIso04_all",
        "MET.pt", "MET.phi", "MET.significance", "MET.covXX", "MET.covXY", "MET.covYY",
        "PV.npvs",
        # all columns added during selection using a ColumnCollection flag
        ColumnCollection.ALL_FROM_SELECTOR,
    },
    "cf.MergeSelectionMasks": {
        "cutflow.*",
    },
    "cf.UniteColumns": {
        "*",
    },
})
```

> -->

Author: done

(custom_retrieval_of_dataset_files_section)=
### Custom retrieval of dataset files

The Columnflow task {py:class}`~columnflow.tasks.external.GetDatasetLFNs` obtains by default the logical file names of the datasets through the [CMS DAS](https://twiki.cern.ch/twiki/bin/view/CMSPublic/WorkBookLocatingDataSamples).
However, this default behaviour can be changed using the auxiliary parameter `cfg.x.get_dataset_lfns`.
If it is not specified or set to `None`, the function :py:meth:`~columnflow.external.GetDatasetLFNs.get_dataset_lfns_dasgoclient` is used.

> Alternatively, you can set it to a custom function with the same keyword arguments as the default.

Author: done

For more information, please consider the documentation of :py:meth:`~columnflow.external.GetDatasetLFNs.get_dataset_lfns_dasgoclient`.
Based on these parameters, the custom function should implement a way to create the list of paths corresponding to this dataset (the paths should not include the path to the remote file system) and return this list.
Two other auxiliary parameters can be changed:

- `get_dataset_lfns_sandbox` provides the sandbox in which the task GetDatasetLFNS will be run and expects therefore a {external+law:py:class}`law.sandbox.base.Sandbox` object, which can be for example obtained through the {py:func}`~columnflow.util.dev_sandbox` function.

> - `get_dataset_lfns_remote_fs` provides the remote file system on which the LFNs for the specific dataset can be found, it expects a function with the `dataset_inst` as a parameter and returning the name of the file system as defined in the law config file.

Author: done

the `expression` of the Variable and storing them in histograms with binning defined
via the `binning` argument of the Variable.

The list of possible keyword arguments can be found in the order documentation for the class
{external+order:py:class}`order.variable.Variable`. The values in the ```expression``` argument can
be either a one-dimensional or a more dimensional array. In this second case the information is
flattened before plotting. It is to be mentioned that
{py:attr}`~columnflow.columnar_util.EMPTY_FLOAT` is a columnflow internal null value and
corresponds to the value ```-9999.0```.

> ### Category

Reviewer 2: This section should also link to the Categories user guide (which is empty for now, but I already started working on that)

Author: done

An example with a selector_steps group is given below.

```python
# selector step groups for conveniently looping over certain steps
# (used in cutflow tasks)
cfg.x.selector_step_groups = {
    "default": ["muon", "jet"],
}
```

With this group defined in the Config object, running over the "muon" and "jet" selector_steps in this order in a cutflow task can done with the argument `--selector-steps default`.

All other parameters for which groups are possible are given below:
```python
# process groups for conveniently looping over certain processs
# (used in wrapper_factory and during plotting)
cfg.x.process_groups = {}

# dataset groups for conveniently looping over certain datasets
# (used in wrapper_factory and during plotting)
cfg.x.dataset_groups = {}

# category groups for conveniently looping over certain categories
# (used during plotting)
cfg.x.category_groups = {}

# variable groups for conveniently looping over certain variables
# (used during plotting)
cfg.x.variable_groups = {}

# shift groups for conveniently looping over certain shifts
# (used during plotting)
cfg.x.shift_groups = {}

> ```

Reviewer 2: There is also the `selector_step_groups`, `calibrator_groups`,  `producer_groups`, `ml_model_groups`

Reviewer 2: That parameter is also rather new, the group resolving is done here: https://github.com/columnflow/columnflow/blob/master/columnflow/tasks/framework/plotting.py#L166

the `expression` of the Variable and storing them in histograms with binning defined
via the `binning` argument of the Variable.

The list of possible keyword arguments can be found in the order documentation for the class
{external+order:py:class}`order.variable.Variable`. The values in the ```expression``` argument can
be either a one-dimensional or a more dimensional array. In this second case the information is
flattened before plotting. It is to be mentioned that
{py:attr}`~columnflow.columnar_util.EMPTY_FLOAT` is a columnflow internal null value and

> corresponds to the value ```-9999.0```.

Author: done

### Category

Categories built to investigate specific parts of the phase-space, for example for plotting.
These objects are described in [the Channel and Category](https://python-order.readthedocs.io/en/latest/quickstart.html#channel-and-category)
part of the Quickstart section of the order documentation. A specificity of columnflow regarding
this class is the fact that the `selection` argument of this object is expected to take the name of
an object of the {py:class}`~columnflow.categorization.Categorizer` class instead of a boolean
expression in a string format.
Adding a category to the Config object is possible through the
{py:func}`~columnflow.config_util.add_category()` method. An example for an inclusive category with

> the Categorizer `cat_incl` defined in the cms_minimal analysis template is given below:

Author: done

```python
add_category(
    cfg,
    id=1,
    name="incl",
    selection="cat_incl",
    label="inclusive",
)

> ```

Author: tried, need to check

> ```

Author: checked

It is recommended to always add an inclusive category with id=1 or name="incl" which is used

> in various places, e.g. for the inclusive cutflow plots and the "empty" selector.

Author: because we wanted to add it to the standard cf functions if I recall correctly and an empty selector would probably be much less interesting without an inclusive category. But I can just delete that

A more detailed description of the usage of categories in columnflow is given in the {ref}`Categories <categories>` section of this documentation.

### Channel

Similarly to categories, Channels are built to investigate specific parts of the phase space and are
described in the [Channel and Category](https://python-order.readthedocs.io/en/latest/quickstart.html#channel-and-category)
part of the Quickstart section of the order documentation. They can be added to the Config object using
{external+order:py:meth}`order.config.Config.add_channel()`.

### Shift

In order to implement systematic variations in the Config object, the
{external+order:py:class}`order.shift.Shift` class can be used. Implementing systematic variations
using shifts can take different forms depending on the kind of systematic variation involved,
therefore a complete section specialized in the description of these implementations is to be found
in (TODO: add link Shift section). Adding a Shift object to the Config object happens through the
{external+order:py:meth}`order.config.Config.add_shift()` function.

> 

Author: done, to check

> 

Author: checked

Often, shifts are related to auxiliary parameters of the Config, like the name of the scale
factors involved, or in the case of corrections accessible through external files, the paths of

> these files.

Author: done

Notes for the Shift section:

-> present examples auxiliary parameters calibrations: cfg.x.jec, cfg.x.muon_sf_names

-> present shift types -> use whole new NanoAODs with specific tune key defined in dataset, or different
calibrations e.g. cfg.x.jec, or weights through producer e.g. scale factors

> -> mention external files

Author: done

## Auxiliary Parameters (optional)

> In principle, the auxiliaries of the Config may contain any kind of variables, but there are

Author: done

several keys with a special meaning for columnflow, for which you would need to respect the expected
format. These are presented below at first, followed by a few examples of the kind of information
you might want to save in the auxiliary part of the config on top of these.

> 

Author: done

### Keep_columns

During the Task {py:class}`~columnflow.tasks.reduction.ReduceEvents` new files in parquet format

> containing all remaining events and objects after the selections are created. If the auxiliary

Author: done

the columns declared explicitely will be kept after the reduction. Actually, several tasks can make
use of such an argument in the Config object for the reduction of their output. Therefore, the
`keep_columns` argument expects a {py:class}`~columnflow.util.DotDict` containing the name of the
tasks (with the `cf.` prefix) for which such a reduction should be applied as keys and the set of
columns to be kept in the output of this task as values.

An example is given below:
```python
cfg.x.keep_columns = DotDict.wrap({
    "cf.ReduceEvents": {
        # general event info
        "run", "luminosityBlock", "event",
        # object info
        "Jet.pt", "Jet.eta", "Jet.phi", "Jet.mass", "Jet.btagDeepFlavB", "Jet.hadronFlavour",
        "Muon.pt", "Muon.eta", "Muon.phi", "Muon.mass", "Muon.pfRelIso04_all",
        "MET.pt", "MET.phi", "MET.significance", "MET.covXX", "MET.covXY", "MET.covYY",
        "PV.npvs",
        # columns added during selection
        "deterministic_seed", "process_id", "mc_weight", "cutflow.*",
    },
    "cf.MergeSelectionMasks": {
        "cutflow.*",
    },
    "cf.UniteColumns": {
        "*",
    },
})

> ```

Author: old version, does not exist anymore, now using ColumnCollection once master branch merged

> ```

Author: done

> For easier handling of the list of columns, the class {py:class}`~columnflow.columnar_util.ColumnCollection` was created.

Author: done

It defines several enumerations containing columns to be kept according to a certain category.
For example, it is possible to keep all the columns created during the SelectEvents task with the enum `ALL_FROM_SELECTOR`.
With the ColumnCollection class, the example above could become:
```python
cfg.x.keep_columns = DotDict.wrap({
    "cf.ReduceEvents": {
        # general event info, mandatory for reading files with coffea
        ColumnCollection.MANDATORY_COFFEA
        # object info
        "Jet.pt", "Jet.eta", "Jet.phi", "Jet.mass", "Jet.btagDeepFlavB", "Jet.hadronFlavour",
        "Muon.pt", "Muon.eta", "Muon.phi", "Muon.mass", "Muon.pfRelIso04_all",
        "MET.pt", "MET.phi", "MET.significance", "MET.covXX", "MET.covXY", "MET.covYY",
        "PV.npvs",
        # all columns added during selection using a ColumnCollection flag
        ColumnCollection.ALL_FROM_SELECTOR,
    },
    "cf.MergeSelectionMasks": {
        "cutflow.*",
    },
    "cf.UniteColumns": {
        "*",
    },
})

> ```

Author: tried, to check once master branch merged

> ```

Author: done

### Custom retrieval of dataset files

The Columnflow task {py:class}`~columnflow.tasks.external.GetDatasetLFNs` obtains by default the
logical file names of the datasets through the
[CMS DAS](https://twiki.cern.ch/twiki/bin/view/CMSPublic/WorkBookLocatingDataSamples).
However, this
default behaviour can be changed using the auxiliary parameter `cfg.x.get_dataset_lfns`,
which must be mapped to either None (triggering the default behaviour and obtaining the LFNs
from DAS) or a custom function with three parameters (`dataset_inst`, the order {external+order:py:class}`order.dataset.Dataset`
object, `shift_inst`, the order {external+order:py:class}`order.shift.Shift` object and
`dataset_key`, the key given in the info dictionary of the dataset when defining it
as shown in the {ref}`Analysis, Campaign and Config section <analysis_campaign_config>`. From these parameters, the custom function should implement

> a way to create the list of LFNs (= the paths to the root files starting from `/store`) corresponding to this dataset and

Author: the exact meaning of dataset_key and of the output might change depending on the experiment yes, so I remove the arguments from the docu and try to formulate it in a more generic way, is it ok like that?

> return this list. Two other auxiliary parameters can be changed, these are called
> `cfg.x.get_dataset_lfns_sandbox` and `cfg.x.get_dataset_lfns_remote_fs`. `get_dataset_lfns_sandbox`
> provides the sandbox in which the task GetDatasetLFNS will be run and expects therefore a
> {external+law:py:class}`law.sandbox.base.Sandbox` object, which can be for example obtained through
> the {py:func}`~columnflow.util.dev_sandbox` function.
> `get_dataset_lfns_remote_fs` provides the remote file system on which the LFNs for the specific dataset can be found, it expects a function with the `dataset_inst` as a parameter and returning the name of the file system as defined in the law config file.

Reviewer: Change to a list?

```suggestion
return this list. Two other auxiliary parameters can be changed

- `get_dataset_lfns_sandbox` provides the sandbox in which the task GetDatasetLFNS will be run and expects therefore a {external+law:py:class}`law.sandbox.base.Sandbox` object, which can be for example obtained through
the {py:func}`~columnflow.util.dev_sandbox` function.
- `get_dataset_lfns_remote_fs` provides the remote file system on which the LFNs for the specific dataset can be found. It expects a function with the `dataset_inst` as a parameter and returning the name of the file system as defined in the law config file.
```

Author: done

            fs=f"wlcg_fs_{cfg.campaign.x.custom['name']}",
        )

        # loop though files and interpret paths as lfns
        return [
            lfn_base.child(basename, type="f").path
            for basename in lfn_base.listdir(pattern="*.root")
        ]

    # define the lfn retrieval function
    cfg.x.get_dataset_lfns = get_dataset_lfns

    # define a custom sandbox
    cfg.x.get_dataset_lfns_sandbox = dev_sandbox("bash::$CF_BASE/sandboxes/cf.sh")

    # define custom remote fs's to look at
    cfg.x.get_dataset_lfns_remote_fs = lambda dataset_inst: f"wlcg_fs_{cfg.campaign.x.custom['name']}"
```

### External_files

If some files from outside columnflow are needed for an analysis, be them local files or online ( and accessible through wget), these can be indicated in the `cfg.x.external_files` auxiliary parameter.
These can then be copied to the columnflow outputs using the {py:class}`~columnflow.tasks.external.BundleExternalFiles` task and used by being required by the object needing them.
The `cfg.x.external_files` parameter expects a {py:class}`~columnflow.util.DotDict` with a string

> explaining the role of the file to be downloaded/copied as key and the link/path as value.

Reviewer: > ... explaining the role of the file to be downloaded/copied as key ...

What do you mean by this? There is no requirement/naming convention at all for these keys, right? If the user so wishes, they could in principle use any kind of key

Author: yes, this would be a best pratices case, I reformulated

It is also possible to give a tuple as value, with the link/path as the first entry of the tuple and a version as a second entry.

> As an example, the `cfg.x.external_files` parameter might look like this, for `json_mirror` the local path of the mirror directory to a specific commit of the [jsonPOG-integration Gitlab](https://gitlab.cern.ch/cms-nanoAOD/jsonpog-integration/-/tree/master) (CMS-specific):

Reviewer: ```suggestion
As an example, the `cfg.x.external_files` parameter might look like this, where `json_mirror` is a local path to a mirror directory of a specific commit of the [jsonPOG-integration Gitlab](https://gitlab.cern.ch/cms-nanoAOD/jsonpog-integration/-/tree/master) (CMS-specific):
```

Author: done

```python
cfg.x.external_files = DotDict.wrap({
    # lumi files
    "lumi": {
        "golden": ("/afs/cern.ch/cms/CAF/CMSCOMM/COMM_DQM/certification/Collisions17/13TeV/Legacy_2017/Cert_294927-306462_13TeV_UL2017_Collisions17_GoldenJSON.txt", "v1"),  # noqa
        "normtag": ("/afs/cern.ch/user/l/lumipro/public/Normtags/normtag_PHYSICS.json", "v1"),
    },

    # muon scale factors
    "muon_sf": (f"{json_mirror}/POG/MUO/2017_UL/muon_Z.json.gz", "v1"),

    # some example of a ML training dataset to show an example with URL
    "vehicules_dataset_fastai": ("https://github.com/arunoda/fastai-courses/releases/download/fastai-vehicles-dataset/fastai-vehicles.tgz", "v1"),  # noqa
})

> ```

Reviewer: Import from example analysis?

Author: does not exist in this form, no example of an path with URL... I could do two examples if you really want to use the template here

> An example of usage of the `muon_sf`, including the requirement of the BundleExternalFiles task is given in the {py:mod}`~columnflow.production.cms.muon` file, implementing the {py:class}`~columnflow.production.cms.muon.muon_weights` Producer.

Reviewer: I think simply the link to the module should be enough

```suggestion
An example of usage of the `muon_sf`, including the requirement of the BundleExternalFiles task, is given in the {py:class}`~columnflow.production.cms.muon.muon_weights` Producer.
```

Author: done

TODO: put the following in drop down menu

Showing how to require the BundleExternalFiles task to have run in the example of the muon weights Producer linked above:
```python
@muon_weights.requires
def muon_weights_requires(self: Producer, reqs: dict) -> None:
    if "external_files" in reqs:
        return

    from columnflow.tasks.external import BundleExternalFiles
    reqs["external_files"] = BundleExternalFiles.req(self.task)

> ```

Reviewer: Include directly from file

Author: checked

### Luminosity

The luminosity, needed for some normalizations and for the labels in the standard columnflow plots, needs to be given in the auxiliary arguments `cfg.x.luminosity` as an object of the {external+scinum:py:class}`scinum.Number` class, such that for example the `nominal` parameter exists.
An example for a CMS luminosity of 2017 with uncertainty sources given as relative errors is given below.

```python
from scinum import Number
cfg.x.luminosity = Number(41480, {
    "lumi_13TeV_2017": 0.02j,
    "lumi_13TeV_1718": 0.006j,
    "lumi_13TeV_correlated": 0.009j,
})
```

### Defaults

Default values can be given for several command line parameters in columnflow, using the `cfg.x.default_{parameter}` entry in the Config object.

> The expected format is a single string containing the name of the object to be used as default for parameters accepting only one argument and a tuple for parameters accepting several arguments.

Reviewer: > The expected format is a single string containing the name of the object to be used as default for parameters accepting only one argument and a tuple for parameters accepting several arguments.

I have read this sentence a couple of times now, and I'm still not sure what you would like to say. Maybe something like

```suggestion
The expected format is either 

- a single string containing the name of the object to be used as default for parameters accepting only one argument or
- a tuple for parameters accepting several arguments.
```

The command-line arguments supporting a default value from the Config object are given in the cms_minimal example of the analysis_templates and shown again below:

```python
cfg.x.default_calibrator = "example"
cfg.x.default_selector = "example"
cfg.x.default_producer = "example"
cfg.x.default_ml_model = None
cfg.x.default_inference_model = "example"
cfg.x.default_categories = ("incl",)
cfg.x.default_variables = ("n_jet", "jet1_pt")
cfg.x.default_weight_producer = "example"

> ```

Reviewer: Would rather import this from the example analysis directly

Author: I added the `cfg.x.default_weight_producer = "example"` inside, which doesn't exist in the current master branch (still in pull request) and was not added to the template in the corresponding pull request, should it just be ignored for now?

Author: done, removed

### Groups

It is also possible to create groups, which allow to conveniently loop over certain command-line parameters.
This is done with the `cfg.x.{parameter}_group` arguments.
The expected format of the group is a dictionary containing the custom name of the groups as keys and the list of the parameter values as values.
The name of the group can then be given as command-line argument instead of the single values.
An example with a selector_steps group is given below.

```python
# selector step groups for conveniently looping over certain steps
# (used in cutflow tasks)
cfg.x.selector_step_groups = {
    "default": ["muon", "jet"],
}

> ```

Reviewer: Import?

Author: checked

With this group defined in the Config object, running over the "muon" and "jet" selector_steps in this order in a cutflow task can done with the argument `--selector-steps default`.

All parameters for which groups are possible are given below:

> ```python

Reviewer: Import whole block?

Author: not all inside the template, so this one would be a bad idea

Author: added to template

cfg.x.variable_groups = {}

# shift groups for conveniently looping over certain shifts
# (used during plotting)
cfg.x.shift_groups = {}

# general_settings groups for conveniently looping over different values for the general-settings parameter
# (used during plotting)
cfg.x.general_settings_groups = {}

# process_settings groups for conveniently looping over different values for the process-settings parameter
# (used during plotting)
cfg.x.process_settings_groups = {}

# variable_settings groups for conveniently looping over different values for the variable-settings parameter
# (used during plotting)
cfg.x.variable_settings_groups = {}

# custom_style_config groups for conveniently looping over certain style configs
# (used during plotting)
cfg.x.custom_style_config_groups = {}

# calibrator groups for conveniently looping over certain calibrators
# (used during calibration)
cfg.x.calibrator_groups = {}

# producer groups for conveniently looping over certain producers
# (used during the ProduceColumns task)
cfg.x.producer_groups = {}

# ml_model groups for conveniently looping over certain ml_models
# (used during the machine learning tasks)
cfg.x.ml_model_groups = {}
```

> ### Reduced Files size

Author: done

The target size of the files in MB after the {py:class}`~columnflow.tasks.reduction.MergeReducedEvents` task can be set in the Config object with the `cfg.x.reduced_file_size` argument. A float number corresponding to the size in MB is expected. If nothing is set, the default value implemented in columnflow will be used (512MB at the time of writing). An example is given below.

```python
# target file size after MergeReducedEvents in MB
cfg.x.reduced_file_size = 512.0

> ```

Author: done, is that ok like that?

### Object-specific variables

> Other than the variables mentioned above, several might be needed for specific Producers for example, but these won't be discussed here as they are not general parameters.

Reviewer: ```suggestion
Other than the variables mentioned above, several might be needed for specific Producers for example.
These won't be discussed here as they are not general parameters.
```

Author: done

Hence, we invite the users to check which Config entries are needed for each {py:class}`~columnflow.calibration.Calibrator`,
{py:class}`~columnflow.selection.Selector` and {py:class}`~columnflow.production.Producer` and in general each CMS-specific object (=objects in the cms-subfolders) they want to use.
Since the {py:class}`~columnflow.production.cms.muon.muon_weights` Producer was already mentioned above, we will remind users here that the `cfg.x.muon_sf_names` Config entry is needed for this Producer to run, as indicated in the docstring of the Producer.

> As for the CMS-specific objects, an example could be the task {py:class}`~columnflow.tasks.cms.external.CreatePileupWeights`, which requires additional entries in the Config objects, for example the minimum bias cross sections `cfg.x.minbias_xs` and the `pu` entry to the external files.

Reviewer: You already made it clear that this list here requires additional information from the config

```suggestion
As for the CMS-specific objects, an example could be the task {py:class}`~columnflow.tasks.cms.external.CreatePileupWeights`, which requires for example the minimum bias cross sections `cfg.x.minbias_xs` and the `pu` entry in the external files.
```
